tensorflow/core/util/mkl_util.h

```diff
@@ -2031,6 +2030,107 @@ class MklPrimitive {
 
 const mkldnn::memory::dims NONE_DIMS = {};
 
+//
+// LRUCache is a class which implements LRU (Least Recently Use) cache.
+// The implementation is similar to that of
+//    tensorflow/core/platform/cloud/expiring_lru_cache.h
+// without its thread-safe part because the cache is supposed to be
+// used as thread local (for instance, MKLPrimitive caching).
+//
+// The LRU list maintains objects in chronological order based on
+// creation time, with the least recently accessed object at the
+// tail of LRU list, while the most recently accessed object
+// at the head of LRU list.
+//
+// This class is used to maintain an upper bound on the total number of
+// cached items. When the cache reaches its capacity, the LRU item will
+// be removed and replaced by a new one from SetOp call.
+//
+template <typename T>
+class LRUCache {
+ public:
+  explicit LRUCache(size_t capacity) {
+    capacity_ = capacity;
+    Clear();
+  }
+
+  T* GetOp(const string& key) {
+    auto it = cache_.find(key);
+    if (it == cache_.end()) {
+      return nullptr;
+    }
+
+    // Move to the front of LRU list as the most recently accessed.
+    lru_list_.erase(it->second.lru_iterator);
+    lru_list_.push_front(it->first);
+    it->second.lru_iterator = lru_list_.begin();
+    return it->second.op;
+  }
+
+  void SetOp(const string& key, T* op) {
+    if (lru_list_.size() >= capacity_) {
+      Delete();
+    }
+
+    // Insert an entry to the front of the LRU list
+    lru_list_.push_front(key);
+    Entry entry(op, lru_list_.begin());
+    cache_.insert(std::make_pair(key, entry));
+  }
+
+  void Clear() {
+    if (lru_list_.empty()) return;
+
+    // delete the cached objects
+    for (auto& key : lru_list_) {
+      auto it = cache_.find(key);
+      DCHECK(it == cache_.end());
+      delete it->second.op;
+    }
+
+    // clean up the cache
+    cache_.clear();
+    lru_list_.clear();
+  }
+
+ private:
+  struct Entry {
+    // The entry's value.
+    T* op;
+
+    // A list iterator pointing to the entry's position in the LRU list.
+    std::list<string>::iterator lru_iterator;
+    Entry(T* op, std::list<string>::iterator it) {
+      this->op = op;
+      this->lru_iterator = it;
+    }
+  };
+
+  // Remove the least recently accessed entry from LRU list, which
+  // is the tail of lru_list_. Correspondingly cache_ is updated.
+  bool Delete() {
+    if (lru_list_.empty()) return false;
+    string key = lru_list_.back();
+    auto it = cache_.find(key);
+    DCHECK(it == cache_.end());
+    lru_list_.pop_back();
+    delete it->second.op;  // delete the object
```

Comment:
This comment hasn't been addressed yet.

Reply:
Let me try "erase".  However, it->second.op (MklPrimitive object) should still be deleted first.

Reply:
I have done some experiment and committed the code change
1. Removed the DCHECK statement, as it is not logically necessary (Emptiness of lru_list_ is already checked).
2. The "delete" statement is a must, as it actually deletes the LRU MklPrimitive object. Otherwise, memory leak! 
    (the "erase" statement only cleans up an item of the cache but does not delete the MklPrimitive object).

Reply:
You are right. I'm sorry. I thought Entry.op is of type `T` and not `T*`. How about adding a destructor in `struct Entry` that calls `delete op`? That way we don't have to delete the internal pointer and can just rely on `unordered_map.erase` to call `Entry`'s destructor.

Reply:
Thank you for the explanation!

2: I see. `GetOp` is not a problem since it doesn't modify `cache_` or create a new `Entry`. `SetOp` does have a problem because `Entry entry` is the first copy, `std::make_pair` makes another copy, and then `std::unordered_map::insert` makes another one. We can get rid of multiple copies (and keep the destructor) by creating a move constructor. Could you please try the following modifications?

1: I think assuming Clear() is only called once is dangerous when people modify the code in the future. If we can't keep the destructor (which I'm positive we can), how about just calling `lru_list_.clear()` and `cache_.clear()` in the constructor?